k4FWCore/src/PodioDataSvc.cpp

```diff
@@ -109,14 +126,19 @@ StatusCode PodioDataSvc::i_setRoot(std::string root_path, DataObject* pRootObj)
 }
 
 void PodioDataSvc::endOfRead() {
+  m_eventNum++;
+
+  if (!m_bounds_check_needed) {
+    return;
+  }
+
   StatusCode sc;
-  if (m_eventMax != -1) {
-    if (m_eventNum++ >= m_eventMax - 1) {  // we start counting at 0 thus the -1.
-      info() << "Reached end of file with event " << m_eventMax << endmsg;
-      IEventProcessor* eventProcessor;
-      sc = service("ApplicationMgr", eventProcessor);
-      sc = eventProcessor->stopRun();
-    }
+  // m_eventNum already points to the next event here so check if it is available
+  if (m_numAvailableEvents != -1 && m_eventNum >= m_numAvailableEvents) {
+    info() << "Reached end of file with event " << m_eventNum << "/" << m_numAvailableEvents << endmsg;
```

Comment:
Turns out the -1 was just wrong, I noticed it when I extended this info message.
With the -1 we stop at e.g. `m_eventNum == 99` when 100 events are available. This would be correct because `m_eventNum` starts at 0, if `m_eventNum` would not already point to the next event.

Reply:
I am also not so sure if we can even hit the "m_numAvailableEvents != -1" check, for that we would need a file not containing any events and I think than we don't even reach the endOfRead check

Reply:
Looking through the code, is `m_numAvailableEvents` ever still `-1` here? At least from a quick glance I couldn't see a path where it is not assigned a non-default value(?).

Reply:
Even if there is a file without events, `reader.getEntries` will return 0. Things might get wonky if there is a file with `numeric_limits<unsigned>::max()` events, as that could result in `-1` again ;)

Reply:
> With the -1 we stop at e.g. `m_eventNum == 99` when 100 events are available.

Good this is what I noticed the last day but haven't had time to reproduce it since. I'll have a final look today to see if it behaves correctly always